### ISSUE TITLE
Poor performance and high memory usage

### ISSUE THREAD

**Issue by bandesz** [opsidian/parsley#5]
The parser is quite slow currently and does loads of memory allocations. I did a benchmark on the example JSON parser and compared it to the encoding/json package:

```
BenchmarkParsleyJSON1k-4     	    2000	    651731 ns/op	  795068 B/op	    5668 allocs/op
BenchmarkParsleyJSON10k-4    	     300	   4965598 ns/op	 5009140 B/op	   41798 allocs/op
BenchmarkEncodingJSON1k-4    	   50000	     25124 ns/op	    5954 B/op	     120 allocs/op
BenchmarkEncodingJSON10k-4   	   10000	    201004 ns/op	   44440 B/op	     918 allocs/op
```

Parsley is 20-25x slower.

**Comment by bandesz** [comment]
I rewrote most of the library and made the reader stateless. Although JSON parsing is not the best comparison as this library is never going to shine with linear static parsing, it is still worth to see how much improvement I was able to achieve.

Also I added tests for a 100k big JSON file.

```
BenchmarkParsleyJSON1k-4      	   10000	    155846 ns/op	  131725 B/op	    2392 allocs/op
BenchmarkParsleyJSON10k-4     	    1000	   2129129 ns/op	 3180634 B/op	   18899 allocs/op
BenchmarkParsleyJSON100k-4    	      20	  81724068 ns/op	235580628 B/op	  188878 allocs/op
BenchmarkEncodingJSON1k-4     	   50000	     25145 ns/op	    5923 B/op	     117 allocs/op
BenchmarkEncodingJSON10k-4    	    5000	    206290 ns/op	   44371 B/op	     901 allocs/op
BenchmarkEncodingJSON100k-4   	    1000	   1728021 ns/op	  418964 B/op	    8388 allocs/op
```

**Comment by bandesz** [comment]
The high memory usage was addressed in the 0.8.0 version too:

```
BenchmarkParsleyJSON1k-4      	   20000	     79929 ns/op	   26557 B/op	     570 allocs/op
BenchmarkParsleyJSON10k-4     	    2000	    681273 ns/op	  248000 B/op	    4435 allocs/op
BenchmarkParsleyJSON100k-4    	     200	   6999517 ns/op	 6011530 B/op	   41647 allocs/op
BenchmarkEncodingJSON1k-4     	   50000	     23834 ns/op	    5794 B/op	     113 allocs/op
BenchmarkEncodingJSON10k-4    	   10000	    190699 ns/op	   44140 B/op	     869 allocs/op
BenchmarkEncodingJSON100k-4   	    1000	   1880887 ns/op	  418492 B/op	    8153 allocs/op
```
